### PR TITLE
Propagate Cancel to CompletableFutureTask (kill external process of localcomputationmanager)

### PR DESCRIPTION
**Pull request by jonenst**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
NO

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
BugFix / Feature

**What is the current behavior?** *(You can also link to an open issue here)*
calling cancel(true) on the CompletableFuture returned by the SecurityAnalysis API (probably other APIs too) do not kill the external process (when using an implementation based on LocalComputationManager)

**What is the new behavior (if this is a feature change)?**
kills the external process.
Calling cancel() multiple times now returns true only for the one time that cause a cancel. (it was already the case in some paths of the code, just not all of them)

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
No

### CODE REVIEW

File: computation/src/main/java/com/powsybl/computation/CompletableFutureTask.java [hunk 1]
```diff
@@ -88,4 +88,33 @@ public boolean cancel(boolean mayInterruptIfRunning) {
...
 
+    /**
+     * A CompletableFuture that remembers the original CompletableFutureTask that created it
+     * and propagates the cancel to it. It will also return the return value from the source
+     * for cancel() if it has to cancel it. This means that a second cancel will return false,
+     * unlike for regular CompletableFutures for which all cancels return true (if actually canceled).
+     * @author Jon Harper <jon.harper at rte-france.com>
+     */
+    static class SourceCancelingCompletableFuture<T> extends CompletableFuture<T> {
+        CompletableFuture<?> source;
```
Comment (mathbagu) [inline computation/src/main/java/com/powsybl/computation/CompletableFutureTask.java:99]: This attribute should probably be private and final.

File: computation/src/main/java/com/powsybl/computation/CompletableFutureTask.java [hunk 2]
```diff
@@ -88,4 +88,33 @@ public boolean cancel(boolean mayInterruptIfRunning) {
...
+
+        public SourceCancelingCompletableFuture(CompletableFuture<?> source) {
+            this.source = source;
```
Comment (mathbagu) [inline computation/src/main/java/com/powsybl/computation/CompletableFutureTask.java:102]: Ensure that `source` is not null?

**Review by mathbagu — APPROVED** [review]

**Review by sylvlecl — APPROVED** [review]
Nice finding, this `newIncompleteFuture()` method!

**Comment by geofjamg** [comment]
@jonenst so this PR solved the 3 years old issue https://github.com/powsybl/powsybl-core/issues/667 ?

**Comment by jonenst** [comment]
I completely forgot that I made this issue 3 years ago wow. Switching to java11 finally paid off hehe

**Pull request merged**